### PR TITLE
coverage: Emit the filenames section before encoding per-function mappings

### PR DESCRIPTION
**Pull request by Zalathar**
When embedding coverage information in LLVM IR (and ultimately in the resulting binary), there are two main things that each CGU needs to emit:

- A single `__llvm_covmap` record containing a coverage header, which mostly consists of a list of filenames used by the CGU's coverage mappings.
- Several `__llvm_covfun` records, one for each instrumented function, each of which contains the hash of the list of filenames in the header.

There is a kind of loose cyclic dependency between the two: we need the hash of the file table before we can emit the covfun records, but we need to traverse all of the instrumented functions in order to build the file table.

The existing code works by processing the individual functions first. It lazily adds filenames to the file table, and stores the mostly-complete function records in a temporary list. After this it hashes the file table, emits the header (containing the file table), and then uses the hash to emit all of the function records.

This PR reverses that order: first we traverse all of the functions (without trying to prepare their function records) to build a *complete* file table, and then emit it immediately. At this point we have the file table hash, so we can then proceed to build and emit all of the function records, without needing to store them in an intermediate list.

---

Along the way, this PR makes some necessary changes that are also worthwhile in their own right:
- We split `FunctionCoverage` into distinct collector/finished phases, which neatly avoids some borrow-checker hassles when extracting a function's final expression/mapping data.
- We avoid having to re-sort a function's mappings when preparing the list of filenames that it uses.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs [hunk 1]
```diff
@@ -139,40 +146,64 @@ pub fn finalize(cx: &CodegenCx<'_, '_>) {
...
+        let working_dir: &str = &tcx.sess.opts.working_dir.for_codegen(&tcx.sess).to_string_lossy();
+
+        // Prepare a map from filename symbols to their underlying strings, so
+        // that we can sort by the strings.
+        let mut raw_file_table = FxIndexMap::<Symbol, &str>::default();
+        // Filenames usually come in contiguous runs, so dedup to save work.
+        let all_file_names = all_file_names.into_iter().dedup().collect::<Vec<_>>();
+        for file_name in &all_file_names {
+            raw_file_table.entry(*file_name).or_insert_with(|| file_name.as_str());
+        }
```
Comment (cjgillot) [inline compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs:166]: Can we work directly with the `FxIndexSet<Symbol>`? Or is it too slow to access the interned strings?

File: compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs [hunk 2]
```diff
@@ -165,72 +162,84 @@ impl<'tcx> FunctionCoverage<'tcx> {
...
+    zero_expressions: ZeroExpressions,
+}
+
+impl<'tcx> FunctionCoverage<'tcx> {
+    fn from_collector(collector: FunctionCoverageCollector<'tcx>) -> Self {
+        let zero_expressions = collector.identify_zero_expressions();
+        let FunctionCoverageCollector { function_coverage_info, is_used, counters_seen, .. } =
+            collector;
+
+        Self { function_coverage_info, is_used, counters_seen, zero_expressions }
```
Comment (cjgillot) [inline compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:184]: Nit: should this go in `into_finished` and remove this method?

**Comment by Zalathar** [comment]
Addressed review feedback [(diff)](https://github.com/rust-lang/rust/compare/ba3859b77452b896c4606533669a10ff67abe068..c46abe4319e57b79561209d4051c2cd04f5fecd2).

File: compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs (hunk 1 above)
Comment (Zalathar) [reply compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs]: Yeah, that's probably fine.

**Comment by Zalathar** [comment]
Fixed outdated test [(diff)](https://github.com/rust-lang/rust/compare/c46abe4319e57b79561209d4051c2cd04f5fecd2..6af9fef08590fd499370a2f6cbbae9ceacf15336).

@rustbot ready

**Pull request merged**